### PR TITLE
Add new command for performance tesing

### PR DESCRIPTION
**Pull request by shabtaisharon**
for example:
./ds3_java_cli -c performance -b performance_bucket -n 100 -s 100 --http
-n 100 meaning create a job with 100 files
-s 100 meaning each file will be with size of 100MB

### CODE REVIEW

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/Arguments.java [hunk 1]
```diff
@@ -48,8 +48,9 @@
     private final String[] args;
     private String objectName;
     private String proxy;
-    private int start;
-    private int end;
+    //TODO re-add these calls when we have support for partial file gets in the helper functions
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/Arguments.java]: I would just outright remove these.  I don't think we have any plans to add partial gets back at any time.  And they are supported in the helper functions now.

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/Arguments.java [hunk 2]
```diff
@@ -87,10 +90,11 @@ public Arguments(final String[] args) throws BadArgumentException, ParseExceptio
         prefix.setArgName("prefix");
         final Option proxy = new Option("x", true, "The URL of the proxy server to use or have \"http_proxy\" set as an environment variable");
         proxy.setArgName("proxy");
-        final Option start = new Option("s", true, "The starting byte for a get_object command");
-        start.setArgName("start");
-        final Option end = new Option("n", true, "The ending byte for a get_object command");
-        end.setArgName("end");
+        //TODO re-add these calls when we have support for partial file gets in the helper functions
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/Arguments.java]: Same comment about removing these.

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/command/Performance.java [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+        final Iterable<Contents> objects = helpers.listObjects(bucketName);
+        for(final Contents contents : objects) {
+            client.deleteObject(new DeleteObjectRequest(bucketName, contents.getKey()));
+        }
+
+        client.deleteBucket(new DeleteBucketRequest(bucketName));
+    }
+}
+
+class MemoryObjectChannelBuilder implements Ds3ClientHelpers.ObjectChannelBuilder {
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/command/Performance.java]: This is not how you typically do inner classes like this in Java.  They should actually be enclosed inside of the Performance class and then made private.

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/command/Performance.java [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+import com.spectralogic.ds3client.models.bulk.Ds3Object;
+import org.apache.commons.cli.MissingOptionException;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.SeekableByteChannel;
+import java.security.SignatureException;
+import java.util.List;
+
+public class Performance extends CliCommand<PerformanceResult>{
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/command/Performance.java]: I would rename this PerformanceCommand

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/command/Performance.java [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+            helpers.ensureBucketExists(this.bucketName);
+
+            final List<Ds3Object> objList = Lists.newArrayList();
+            for (int i = 0; i<numberOfFiles ; i++) {
+                final long testFileSize = sizeOfFiles * 1024L * 1024L;
+                final Ds3Object obj = new Ds3Object("file_" + i, testFileSize);
+                objList.add(obj);
+            }
+
+            /**** PUT ****/
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/command/Performance.java]: Split put and get into separate methods.

**Comment by rpmoore** [comment]
It doesn't look like there is a json view for the PerformanceCommand.  With the way the command is written it would be difficult to implement, so I think it's fine if we print a warning that json output is not supported with the performance mode if they specify it.

**Comment by shabtaisharon** [comment]
committed all the changes

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/command/PerformanceCommand.java [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+    @Override
+    public PerformanceResult call() throws Exception {
+        final Ds3ClientHelpers helpers = Ds3ClientHelpers.wrap(getClient());
+        final int numberOfFiles = Integer.valueOf(this.numberOfFiles);
+        final long sizeOfFiles = Integer.valueOf(this.sizeOfFiles);
+
+        try {
+            helpers.ensureBucketExists(this.bucketName);
+
+            final List<Ds3Object> objList = Lists.newArrayList();
```
Comment (rpmoore) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/command/PerformanceCommand.java]: I would also move this into it's own function.

**Comment by rpmoore** [comment]
What is the output when you try to run the performance command with the json output specified.

File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/command/PerformanceCommand.java (hunk 6 above)
Comment (shabtaisharon) [inline ds3_java_cli/src/main/java/com/spectralogic/ds3cli/command/PerformanceCommand.java]: done

**Comment by shabtaisharon** [comment]
Put Statistics: (1/1 MB), files (1/1), Time (1.604 sec), Mbps (0.623), Highest Mbps (0.623 sec)
Get Statistics: (1/1 MB), files (1/1), Time (1.060 sec), Mbps (0.943), Highest Mbps (0.943 sec)
Json output is not supported with the performance command

**Comment by shabtaisharon** [comment]
maybe an empty string is better here?

**Comment by rpmoore** [comment]
We probably want to error out without running.  So in the init for the command check to see if the output is json, and if it is through an exception there.  Then we don't need the json view for it at all.

**Comment by shabtaisharon** [comment]
will do that boss

**Comment by shabtaisharon** [comment]
this is better?
{
  "Meta" : {
    "Date" : "2016-01-08T21:54:10.561Z"
  },
  "Status" : "ERROR",
  "Message" : "Json output is not supported with the performance command"
}

Process finished with exit code 1

**Comment by rpmoore** [comment]
Perfect

**Pull request merged**